[CA-1580] Refactor / serialize policy updates in create workspace

andy7i opened the pull request:
QA-1588

Run policy updates serially in create workspace to prevent akka subscription timeout.

---

I'm seeing this error in the call to syncPolicyToGoogle:

```
500 Internal Server Error entity: {"causes":[],"exceptionClass":"java.util.concurrent.TimeoutException","message":"Response entity was not subscribed after 1 second. Make sure to read the response `entity` body or call `entity.discardBytes()` on it -- in case you deal with `HttpResponse`, use the shortcut `response.discardEntityBytes()`. POST /api/google/v1/resource/workspace/58410ad7-4632-4fbb-8953-19abbe0ae1c5/owner/sync Empty -> 200 OK Chunked
```

the response entity is read immediately after the request is made: https://github.com/broadinstitute/rawls/blob/5a5582ec9ba0ed1cb751f2a0e20523e67abda[…]/scala/org/broadinstitute/dsde/rawls/util/HttpClientUtils.scala

 

One option is to increase the akka subscription timeout to prevent this issue: Substream Source cannot be materialized more than once when reading from s30 - NEW. I tested this out and while it does fix this issue, I'm a bit hesitant to do this as it is an app-wide setting. 

The other is to serially make policy sync calls. This may make createWorkspace even a little longer, but could also fix QA-1571: Sam call to HttpMethod(DELETE) /api/resources/v1/billing-project/abc failed IN PROGRESS as well.





- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

andy7i committed:
not sure if this is the right way to run serially.

also not sure if using the implicit CS is the right thing to do here.

```
  implicit lazy val contextShift: ContextShift[IO] = cats.effect.IO.contextShift(executionContext)
```

andy7i committed:
add type declaration

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -77,7 +77,8 @@ object WorkspaceService {
                  (userInfo: UserInfo)
-                 (implicit system: ActorSystem, materializer: Materializer, executionContext: ExecutionContext): WorkspaceService = {
+                 (implicit system: ActorSystem, materializer: Materializer, executionContext: ExecutionContext,
+                  cs: ContextShift[IO]): WorkspaceService = {
```
also not sure if using the implicit CS is the right thing to do here, or if we should create a new ContextShift

```
  implicit lazy val contextShift: ContextShift[IO] = cats.effect.IO.contextShift(executionContext)
```

gpolumbo-broad replied:
I thought you said you talked with Qi and were just going to use the implicit here?  Or am I misunderstanding?

andy7i commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2178,22 +2179,24 @@ class WorkspaceService(protected val userInfo: UserInfo,
-      Future.traverse(policyEmailsByName.keys) { policyName =>
+      // This used to be Future.traverse but that runs all the futures eagerly and concurrently. This led to some perf
+      // issues, hitting the akka subscription timeout. We switched to use cats traverse to make these calls serially.
+      policyEmailsByName.keys.toList.traverse { policyName =>
```
not sure if this is the right/best way to run serially

andy7i reviewed (commented):
looking for feedback on this approach

ehigham commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2178,22 +2179,24 @@ class WorkspaceService(protected val userInfo: UserInfo,
+      // issues, hitting the akka subscription timeout. We switched to use cats traverse to make these calls serially.
+      policyEmailsByName.keys.toList.traverse { policyName =>
         if (policyName == SamWorkspacePolicyNames.projectOwner && workspaceRequest.authorizationDomain.getOrElse(Set.empty).isEmpty) {
           // when there isn't an auth domain, we will use the billing project admin policy email directly on workspace
           // resources instead of synching an extra group. This helps to keep the number of google groups a user is in below
           // the limit of 2000
-          Future.successful(())
+          IO.fromFuture(IO(Future.successful(())))
         } else if (WorkspaceAccessLevels.withPolicyName(policyName.value).isDefined || policyName == SamWorkspacePolicyNames.canCompute) {
           // only sync policies that have corresponding WorkspaceAccessLevels to google because only those are
           // granted bucket access (and thus need a google group)
-          traceWithParent(s"syncPolicy-${policyName}", s1)(_ => samDAO.syncPolicyToGoogle(SamResourceTypeNames.workspace, workspaceId, policyName))
+          IO.fromFuture(IO(traceWithParent(s"syncPolicy-${policyName}", s1)(_ => samDAO.syncPolicyToGoogle(SamResourceTypeNames.workspace, workspaceId, policyName))))
         } else {
-          Future.successful(())
+          IO.fromFuture(IO(Future.successful(())))
         }
-      }
+      }.unsafeToFuture()
```
I think all this can be simplified into one [Applicative[Future].whenA](https://typelevel.org/cats/api/cats/Applicative.html#whenA[A](cond:Boolean)(f:=%3EF[A]):F[Unit]) call (or whatever the logic is, it's too confusing)
```scala
// why do we not have these predicates???
def isProjectOwner(policy: SamResourcePolicyName): Boolean = 
  policy == SamWorkspacePolicyNames.projectOwner

def isCanCompute(policy: SamResourcePolicyName): Boolean = 
  policy == SamWorkspacePolicyNames.canCompute

def hasAuthorizationDomain(request: WorkspaceRequest): Boolean =
  request.authorizationDomain match {
    case Some(xs) => !xs.isEmpty
    case None => False
  }

...

Applicative[Future].whenA(
  !isProjectOwner(policyName) && 
  hasAuthorizationDomain(workspaceRequest) &&
  (WorkspaceAccessLevels.withPolicyName(policyName.value).isDefined ||  isCanCompute(policyName)),
  traceWithParent(s"syncPolicy-${policyName}", s1)(_ => 
    samDAO.syncPolicyToGoogle(SamResourceTypeNames.workspace, workspaceId, policyName)
  )
)
```

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2005,7 +2006,7 @@ class WorkspaceService(protected val userInfo: UserInfo,
       googleProjectLabels = gcsDAO.labelSafeMap(Map("workspaceNamespace" -> workspaceName.namespace, "workspaceName" -> workspaceName.name, "workspaceId" -> workspaceId), "")
       googleProjectName = gcsDAO.googleProjectNameSafeString(s"${workspaceName.namespace}--${workspaceName.name}")
-      _ = logger.info(s"Setting up project in ${googleProjectId} cloud resource manager.")
+      _ = logger.info(s"Setting up project ${googleProjectId} in cloud resource manager.")
```
```suggestion
      _ = logger.info(s"Setting up google project ${googleProjectId} in cloud resource manager.")
```

gpolumbo-broad reviewed (approved):
LGTM so long as we can get a double check on the `ContextShift` from an `IO` expert

andy7i committed:
Update core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

Co-authored-by: Greg Polumbo <36171087+gpolumbo-broad@users.noreply.github.com>

andy7i committed:
pr feedback -- using Applicative[Future].whenA

andy7i committed:
fixed faulty condition

andy7i commented:
confirmed with @Qi77Qi that we should use the default ContextShift.

ehigham reviewed (approved):
Looks great. Thank you for making that change. I wonder if those helper functions and predicates would be better placed elsewhere. For example, those predicates that test policy names could they live in the same namespace as where the policy names are defined? That might enable better reuse.

ehigham commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2178,23 +2180,36 @@ class WorkspaceService(protected val userInfo: UserInfo,
-        } else {
-          Future.successful(())
-        }
+  private def syncPolicies(workspaceId: String, policyEmailsByName: Map[SamResourcePolicyName, WorkbenchEmail], workspaceRequest: WorkspaceRequest, parentSpan: Span): Future[List[Any]] = {
```
If you want to, you can simplify the return type that you. Just `Future[Unit]` using `traverse_` instead of `traverse`

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2178,23 +2180,36 @@ class WorkspaceService(protected val userInfo: UserInfo,
+      // This used to be Future.traverse but that runs all the futures eagerly and concurrently. This led to a perf
+      // issue with hitting the akka subscription timeout. We switched to use cats traverse to make these calls serially.
+      policyEmailsByName.keys.toList.traverse { policyName =>
+        Applicative[Future].whenA(
```
I thought you want this sync call to happen sequentially? in that case, you do want to convert these `Future` into `IO`, and run `traverse` (or `traverse_` if you're returning `IO[Unit]`) on`IO` instead of `Future`

andy7i committed:
PR feedback - run serially for real

andy7i committed:
PR feedback - move predicates to case class

andy7i commented:
update: ran the workspaces test against perf 3 times and the create workspace tests all passed. Kibana shows 0 instances of the error akka subscription timeout error from these runs.

gpolumbo-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:
```diff
@@ -2178,23 +2180,24 @@ class WorkspaceService(protected val userInfo: UserInfo,
-      }
-    )
+          !(policyName.isProjectOwner && !workspaceRequest.hasAuthorizationDomain) &&
+            (WorkspaceAccessLevels.withPolicyName(policyName.value).isDefined || policyName.isCanCompute)) (
```
I'm not thrilled about this logic statement, it took me a long time to grok what this is trying to say which has me worried if we have it right in the first place and has me worried that this will be fragile/difficult to maintain.

Long story short, I wonder if we create a new enumeration/list of policy names that should be synced.  Something like: 

`// Seq[String] of the policy names on a Workspace that need to be sync'd with their corresponding policy group on GSuite
val policyNamesThatRequireSync: Seq[String] = WorkspaceAccessLevels.all.map(toString) :+ SamWorkspacePolicyNames.canCompute.toString()`

(might be able to pretty that up too)

And then if we apply DeMorgan's to the first predicate, I think the logic can be expressed more simply as:

`(workspaceRequest.hasAuthorizationDomain || !policyName.isProjectOwner) && policyNamesThatRequireSync.contains(policyName.value)`

Which I think boils down do the idea that if the policy is one a "sync'able" policy and the Workspace has an AuthDomain, then go ahead and sync it.  If the Workspace does not have an AuthDomain, then skip the `Project Owner` policy.  Which has me thinking instead we go to:

`// Seq[String] of the policy names on a Workspace that need to be sync'd with their corresponding policy group on GSuite.  If there is not an AuthDomain, then we do not need to sync the Project Owner policy
def policyNamesThatRequireSync(withAuthDomain: Boolean): Seq[String] = { 
    val policyNames = WorkspaceAccessLevels.all.map(toString) :+ SamWorkspacePolicyNames.canCompute.toString()
    if (withAuthDomain) policyNames else policyNames.filterNot(_ == WorkspaceAccessLevels.ProjectOwner.value)
}`

So finally, I think this logical expression I'm commenting on can be boiled down to:

`policyNamesThatRequireSync(workspaceRequest.hasAuthorizationDomain).contains(policyName.value)`

Sorry this is such a long comment, it took me several tries to think this all through and to try to express this logic more clearly.  I really really hope I've got it right, I'm not convinced I do, which is exactly the reason why I wanted to try to simplify the expression.

andy7i committed:
PR feedback -- refactoring

note to self:
```

  //          !(policyName.isProjectOwner && !workspaceRequest.hasAuthorizationDomain) &&
  // true true  true
  // true false false
  // false true true
  // false false  true
  //          (!policyName.isProjectOwner || workspaceRequest.hasAuthorizationDomain) &&
  // true true true
  // true false false
  // false true true
  // false false  true
  ```
  tested demorgans

andy7i commented:
putting down this refactor and tracking it as part of CA-1580 since it is not urgent and not a team priority
